### PR TITLE
update gosnmp and add support for opaque float/double

### PR DESCRIPTION
**Pull request by mschoch**
updated vendored version of gosnmp to latest, which now supports
returning values from opaque float and double

updated generator to recognize opaque float/double and configure
it to work as a gauge

updated collector to extract these values correctly

@brian-brazil

### CODE REVIEW

File: generator/tree.go [hunk 1]
```diff
@@ -95,12 +95,24 @@ func prepareTree(nodes *Node) map[string]*Node {
 		}
 	})
 
+	// Promote Opaque Float/Double textual convention to type
+	walkNode(nodes, func(n *Node) {
+		if n.Type == "OPAQUE" {
+			if strings.EqualFold(n.TextualConvention, "float") {
```
Comment (brian-brazil) [inline generator/tree.go]: We do case-sensitive comparisons everywhere else

File: generator/tree.go [hunk 2]
```diff
@@ -95,12 +95,24 @@ func prepareTree(nodes *Node) map[string]*Node {
...
+				n.Type = "FLOAT"
```
Comment (brian-brazil) [inline generator/tree.go]: `Float`, as that's what's in the MIB

File: generator/tree_test.go [hunk 3]
```diff
@@ -782,6 +791,42 @@ func TestGenerateConfigModule(t *testing.T) {
 				},
 			},
 		},
+		// Opaque Float becomes gauge
```
Comment (brian-brazil) [inline generator/tree_test.go]: These go up with the metric type case

File: collector.go [hunk 4]
```diff
@@ -172,6 +172,10 @@ func getPduValue(pdu *gosnmp.SnmpPDU) float64 {
 	switch pdu.Type {
```
Comment (brian-brazil) [inline collector.go]: pduValueAsString also needs updating.

**Comment by brian-brazil** [comment]
FORMAT.md will also need updating.

**Comment by mschoch** [comment]
What is your preferred way for subsequent changes, additional commits to this branch? Or force push amended commit?

**Comment by brian-brazil** [comment]
Additional commits tend to be easier to review.

**Comment by mschoch** [comment]
Can you clarify how FORMAT.md is updated/used?  The text at the top indicates that it is automatically generated, but I don't see a description of how that is done.

**Comment by brian-brazil** [comment]
FORMAT.md is the documentation for what snmp.yml is. You're adding new types, so you need to add them in there by hand.

**Comment by mschoch** [comment]
Thanks, I'm still a bit confused, as this change results in Float and Double being generated as a gauge in the resulting snmp.yml.  My understanding was that at the prometheus level everything had to turn into gauge or counter anyway.  But I do see other types documented in this file, so maybe my understanding and part of this PR is wrong.

**Comment by mschoch** [comment]
I think maybe I get it now, perhaps instead of turning Float and Double into a gauge, they should stay as their own type in the snmp.yml.  Then that additional type information might let you format or otherwise handle it a bit differently inside the snmp_exporter.  Then at the level of prometheus scraping it, they all become gauages or counters anyway.  Does this sound correct?

**Comment by brian-brazil** [comment]
My question would be how does gosnmp know that a value is a float? That information is only available from the MIB afaik, which gosnmp does not have access to.

**Comment by mschoch** [comment]
What I learned when I hacked the gosnmp portion was that the opaque type values are not just arbitrary bytes, but actually double-wrapped ASN.1 values.  The outer wrapper has type opaque (0x44) and the inner value has types 0x78 (float) or 0x78 (double).

It appears to me that opaque was included initially for extensibility, but it was under-specified and became confusing.  Later, they deprecated it, but there are a few types like Float, Double and possibly a few others that are still used.

- https://tools.ietf.org/html/draft-perkins-opaque-01
- https://tools.ietf.org/html/draft-perkins-float-00

Note, both of those are expired drafts, so it's possible this is not formal, but just a convention people happen to use (but following this did allow me to successfully parse the data I found in the real world).

As for your original question, both the MIB textual convention and type inside wrapped opaque value give us indication the value should be interpreted as float/double.

**Comment by brian-brazil** [comment]
To be consistent let's work off the textualconvention like we do elsewhere then. This will avoid being over-eager in detecting such values.

**Comment by mschoch** [comment]
Thanks for helping work through this, I'm still not exactly clear how you'd like it to proceed.  This PR does have the generator detect Float/Double based on the textual convention in the MIB.  I've changed it now to keep Float/Double as types in that snmp.yml.  I've documented these types in FORMAT.md.  However, I then also had to add some code to properly render these as a guage when scraped by prometheus.  I've also tried to incorporate your other feedback.  Hopefully this gets us a little closer.

File: generator/tree.go [hunk 5]
```diff
@@ -95,6 +95,15 @@ func prepareTree(nodes *Node) map[string]*Node {
 		}
 	})
 
+	// Promote Opaque Float/Double textual convention to type
```
Comment (brian-brazil) [inline generator/tree.go]: Comments should end in a full stop.

File: generator/tree.go [hunk 6]
```diff
@@ -95,6 +95,15 @@ func prepareTree(nodes *Node) map[string]*Node {
...
+	walkNode(nodes, func(n *Node) {
+		if n.Type == "OPAQUE" {
```
Comment (brian-brazil) [inline generator/tree.go]: I'd expect that checking the textual convention is enough

File: generator/tree.go [hunk 7]
```diff
@@ -113,6 +122,10 @@ func metricType(t string) (string, bool) {
 		return "InetAddress", true
 	case "PhysAddress48", "DisplayString":
 		return t, true
+	case "Float":
```
Comment (brian-brazil) [inline generator/tree.go]: You can combine this into the previous case

**Comment by mschoch** [comment]
Pushed commit addressing most recent review comments.

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**